### ISSUE TITLE
False positive on recursive dataclass + union + generic alias

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#2717]
This code:
```python
from dataclasses import dataclass
from typing import Generic, Sequence, TypeVar, Union


class Node:
    pass


NodeT = TypeVar("NodeT", bound=Node)


class WithTrailingComma(Generic[NodeT]):
    node: NodeT


@dataclass
class SimpleJoinedTable(Node):
    left: "TableReference"


class SubqueryFactor(Node):
    pass


@dataclass
class TableReferenceList(Node):
    references: "TableReferences"


TableReference = Union[TableReferenceList, SubqueryFactor]
TableReferences = Sequence[WithTrailingComma[TableReference]]
```
Produces
```
Type "TableReference" cannot be assigned to type variable "NodeT@WithTrailingComma"
  Type "TableReference" is incompatible with bound type "Node" for type variable "NodeT@WithTrailingComma"
    Type "TableReference" cannot be assigned to type "Node"
```
At the last line.

The error goes away if I remove any other part of the program, including the unused `SimpleJoinedTable` class, or if I remove any of the `@dataclass` decorators.

Still on Pylance v2021.12.2.

**Comment by erictraut** [comment]
Supporting this particular combination is difficult given the way that pyright resolves symbols. Normally pyright doesn't resolve the type of a class field until it encounters an expression that accesses that field, but in the case of dataclasses, it needs to resolve the types of the fields when it processes the class because it needs to synthesize an `__init__` method with those types. The same is true for TypedDicts, which involves synthesizing some methods. 

I can think of some potential ways to support this use case if it becomes important enough, but it requires significant changes. Since there is an easy workaround, I don't think it's worth the investment at this time.

To work around this, replace the use of one of the type aliases in the dataclass fields with its expanded form.

I'm going to close this for now, but we might revisit it in the future if others run into this limitation.
